### PR TITLE
Use the new Router interface 

### PR DESCRIPTION
**Pull request by nkcr**
Implementation is not perfect, but ready for a first pass.

### CODE REVIEW

**Comment by nkcr** [comment]
I am having troubles running the DKG scenario (`go test -run TestPedersen_Scenario`) with more than 8 nodes. The error nearly always happen at the very end of the scenario, where the last node ask to decrypt the secret. I suspect there is too many open connections and additional connections can be created. However, it seems that connections are well closed between each stream. Bellow are the diagrams when the error happen and when not:

[graph bad.pdf](https://github.com/dedis/dela/files/5142732/graph.bad.pdf)
[graph ok.pdf](https://github.com/dedis/dela/files/5142733/graph.ok.pdf)
[events bad.pdf](https://github.com/dedis/dela/files/5142734/events.bad.pdf)
[events ok.pdf](https://github.com/dedis/dela/files/5142735/events.ok.pdf)

Also, I am not sure how to handle stream errors and send then back to the user. Using calls to pass messages and streams to monitor the connections could do the trick.

File: mino/minogrpc/server.go [hunk 1]
```diff
@@ -43,7 +48,7 @@ var (
 type overlayServer struct {
 	overlay
 
-	endpoints map[string]Endpoint
+	endpoints *sync.Map // must contain only the type *Endpoint
```
Comment (Gilthoniel) [inline mino/minogrpc/server.go:51]: see comment above

File: mino/minogrpc/server.go [hunk 2]
```diff
@@ -164,42 +170,108 @@ func (o overlayServer) Call(ctx context.Context, msg *Message) (*Message, error)
...
 	if !ok {
 		return xerrors.Errorf("handler '%s' is not registered", uri)
 	}
+	endpoint := itf.(*Endpoint)
 
-	// Listen on the first message, which should be the routing infos
-	msg, err := stream.Recv()
-	if err != nil {
-		return xerrors.Errorf("failed to receive routing message: %v", err)
+	gateway := gatewayFromContext(stream.Context(), o.addrFactory)
```
Comment (Gilthoniel) [inline mino/minogrpc/server.go:187]: I'm not sure why the gateway is in the header.

File: mino/minogrpc/server.go [hunk 3]
```diff
@@ -164,42 +170,108 @@ func (o overlayServer) Call(ctx context.Context, msg *Message) (*Message, error)
...
+	relayCtx := metadata.NewOutgoingContext(stream.Context(), metadata.Pairs(
+		headerURIKey, uri, headerGatewayKey, string(mebuf), headerStreamIDKey,
+		streamID))
+
+	f := func() error {
+		return listenStream(relayCtx, stream, session.sender, gateway)
+	}
+	msg := fmt.Sprintf("failed to listen to our gateway '%s'", gateway)
+
+	go execOrFil(f, session.sender.receiver.errs, msg)
```
Comment (Gilthoniel) [inline mino/minogrpc/server.go:268]: fill ?

File: mino/minogrpc/server.go [hunk 4]
```diff
@@ -208,23 +280,169 @@ func (o overlayServer) Stream(stream Overlay_StreamServer) error {
...
+func sendToRelays(relayCtx context.Context, sender sender, out map[mino.Address]*Envelope) error {
+	for relayAddr, env := range out {
+
+		itf, ok := sender.relays.Load(relayAddr.String())
+		var relay relayer
+
+		if ok {
+			relay = itf.(relayer)
+		} else {
+			conn, err := sender.connFactory.FromAddress(relayAddr)
```
Comment (Gilthoniel) [inline mino/minogrpc/server.go:335]: Reminder to check if a new connection is created every time, or if we reuse them.

File: mino/minogrpc/stream.go [hunk 5]
```diff
@@ -131,6 +105,9 @@ type receiver struct {
 	addressFactory mino.AddressFactory
 	errs           chan error
 	queue          Queue
+
+	ctx    context.Context
```
Comment (Gilthoniel) [inline mino/minogrpc/stream.go:109]: it seems not to be used so I guess you can remove it.

**Review by Gilthoniel — COMMENTED** [review]
Some reminders for when we talk.

File: mino/minogrpc/rpc.go [hunk 6]
```diff
@@ -107,42 +112,41 @@ func (rpc RPC) Stream(ctx context.Context,
...
-	rting, err := rpc.overlay.routingFactory.Make(rpc.overlay.me, players)
+	b := make([]byte, streamIDLen)
+	_, err := rand.Reader.Read(b)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("couldn't generate routing: %v", err)
+		return nil, nil, xerrors.Errorf("failed to create streamID: %v", err)
 	}
 
-	header := metadata.New(map[string]string{headerURIKey: rpc.uri})
+	streamID := fmt.Sprintf("%x-%x-%x-%x-%x",
```
[the remark this answers is not in the document]
Comment (nkcr) [reply mino/minogrpc/rpc.go:121]: Using now https://github.com/rs/xid, which comes from the same source as zerolog.

File: mino/minogrpc/server.go [hunk 7]
```diff
@@ -164,42 +167,116 @@ func (o overlayServer) Call(ctx context.Context, msg *Message) (*Message, error)
...
 
-	err = endpoint.Handler.Stream(sender, receiver)
+	mebuf, err := o.me.MarshalText()
 	if err != nil {
-		return xerrors.Errorf("handler failed to process: %v", err)
+		return xerrors.Errorf("failed to marshal my address: %v", err)
+	}
+
+	relayCtx := metadata.NewOutgoingContext(stream.Context(), metadata.Pairs(
+		headerURIKey, uri, headerGatewayKey, string(mebuf), headerStreamIDKey,
```
Comment (Gilthoniel) [inline mino/minogrpc/server.go:253]: one line per header for readability ?

File: mino/minogrpc/server.go [hunk 8]
```diff
@@ -208,23 +285,234 @@ func (o overlayServer) Stream(stream Overlay_StreamServer) error {
 	return nil
 }
 
-type relayer interface {
+func (s sender) closeRelays() {
```
Comment (Gilthoniel) [inline mino/minogrpc/server.go:288]: Please keep implementation of a struct in a unique file.

File: mino/router/mod.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+	// GetSource returns the source address of the message.
+	GetSource() mino.Address
+
+	// GetDestination returns the destination address of the message.
+	GetDestination() []mino.Address
+
+	// GetMessage returns the message to be transmitted to the application when
+	// the source address is the current node. Message is only deserialized when
+	// needed.
```
Comment (Gilthoniel) [inline mino/router/mod.go:23]: Then return the buffer instead and let mino handles the deserialization, the router doesn't care about it.

File: mino/router/mod.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	// create a map with 3 entries:
+	//
+	//  {A: packet{to: [A]}, B: packet{to: [B]}, C: packet{to: [C]}}.
+	//
+	// For a tree routing, the first message will send a packet with all the
+	// recipients to the root address (A in the following), like:
+	//
+	//	{A: packet{to: [A, B, C]}}
+	//
+	Forward(me mino.Address, data []byte, ctx serde.Context) (map[mino.Address]Packet, error)
```
Comment (Gilthoniel) [inline mino/router/mod.go:47]: it should be `Forward(mino.Address, Packet) (...)`. Again the router does not care about encoding.

Comment (Gilthoniel) [reply mino/router/mod.go:47]: The router can provide the packet factory.

File: mino/router/tree/types/packet.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+// Packet describes a tree routing packet
+//
+// - implements router.Packet
+type Packet struct {
+	Source  mino.Address
+	Dest    []mino.Address
+	Message []byte
+	Seed    int64
```
Comment (Gilthoniel) [inline mino/router/tree/types/packet.go:24]: The fields should be private.

**Comment by Gilthoniel** [comment]
@nkcr I'm going to merge without further review because it should be done by tomorrow. We can always talk about it next week.

**Pull request merged**